### PR TITLE
librustdoc: improve testnames for doctests

### PR DESCRIPTION
**Pull request by Ryman**
Old doctest names

``` bash
test sync::atomic::load_0 ... ok
test sync::atomic::load_0 ... ok
test sync::atomic::load_0 ... ok
test sync::atomic::load_0 ... ok
```

New doctest names

``` bash
test sync::atomic::AtomicBool::load_0 ... ok
test sync::atomic::AtomicIsize::load_0 ... ok
test sync::atomic::AtomicPtr<T>::load_0 ... ok
test sync::atomic::AtomicUsize::load_0 ... ok
```

### CODE REVIEW

File: src/librustdoc/test.rs [hunk 1]
```diff
@@ -422,11 +422,14 @@ impl Collector {
...
-            Some(ref name) => { self.names.push(name.to_string()); true }
-            None => false
-        };
+        let name = item.name.iter()
+                            .filter(|name| !name.is_empty())
+                            .cloned()
+                            .next()
+                            .or_else(|| typename_if_impl(&item));
+
+        let pushed = name.map(|name| self.names.push(name)).is_some();
```
Comment (alexcrichton) [inline src/librustdoc/test.rs]: This is a little bit hard to follow, so perhaps this could use explicit control flow + `if let Some(...) = ...` to match what's going on here? Currently it's somewhat non-idiomatic to call `.iter()` on an option or rely on functions like `map` for side effects.

File: src/librustdoc/test.rs [hunk 2]
```diff
@@ -434,10 +437,43 @@ impl DocFolder for Collector {
...
+                s.replace("&gt;", ">")
+                 .replace("&lt;", "<")
+            } else {
+                s
+            }
+        }
+
+        fn typename_if_impl(item: &clean::Item) -> Option<String> {
+            if let clean::ItemEnum::ImplItem(ref impl_) = item.inner {
+                let path = impl_.for_.to_string();
```
Comment (alexcrichton) [inline src/librustdoc/test.rs]: I think I may be getting lost, but what's this actually calling? I couldn't find `Display for Type` or an explicit `to_string` method, so I'm  not actually sure where this is going...

File: src/librustdoc/clean/mod.rs [hunk 3]
```diff
@@ -2568,8 +2568,12 @@ fn resolve_type(cx: &DocContext,
     debug!("resolve_type({:?},{:?})", path, id);
     let tcx = match cx.tcx_opt() {
         Some(tcx) => tcx,
-        // If we're extracting tests, this return value doesn't matter.
-        None => return Primitive(Bool),
+        // If we're extracting tests, this return value's accuracy is not important, all we
+        // want is a string representation to help people figure out what doctests are failing.
+        None => {
+            let did = DefId::local(DefIndex::from_u32(0));
+            return ResolvedPath { path: path, typarams: None, did: did, is_generic: false }
```
Comment (alexcrichton) [inline src/librustdoc/clean/mod.rs]: I think it's been awhile since I've poked around this, but could you elaborate a bit on why swapping one dummy type for another was needed to get the names to work out?

Also, could you wrap this to 80 chars? (style of rustdoc)

File: src/librustdoc/test.rs (hunk 1 above)
Comment (Ryman) [inline src/librustdoc/test.rs]: Heh, it got a bit out of hand, will do.

File: src/librustdoc/test.rs (hunk 2 above)
Comment (Ryman) [inline src/librustdoc/test.rs]: https://github.com/Ryman/rust/blob/rustdoc_better_testnames/src/librustdoc/html/format.rs#L434

File: src/librustdoc/clean/mod.rs (hunk 3 above)
Comment (Ryman) [inline src/librustdoc/clean/mod.rs]: The majority of types, mainly anything more complicated than a primitive, would become `Primitive(Bool)`, which renders as `bool`, this still skips the type resolve and uses the path we've currently got which is good enough for the test outputs as it should match what the user actually wrote on their impl. (and make things easy to find?)

**Comment by Ryman** [comment]
@alexcrichton Adjusted from feedback

File: src/librustdoc/test.rs (hunk 2 above)
Comment (alexcrichton) [inline src/librustdoc/test.rs]: Oh no it's fine, looks like I'm just abysmal at grep

File: src/librustdoc/test.rs [hunk 4]
```diff
@@ -422,22 +422,59 @@ impl Collector {
...
+                s.replace("&gt;", ">")
+                 .replace("&lt;", "<")
+            } else {
+                s
+            }
+        }
+
+        fn typename_if_impl(item: &clean::Item) -> Option<String> {
+            if let clean::ItemEnum::ImplItem(ref impl_) = item.inner {
+                let path = <clean::Type as ToString>::to_string(&impl_.for_);
```
Comment (alexcrichton) [inline src/librustdoc/test.rs]: Ah yeah it's fine to use `.to_string()` here, I was just getting confused :(

Comment (Ryman) [inline src/librustdoc/test.rs]: Just pushed adding`.to_string()` again :)
